File: Copy List with Random Pointer - Leetcode 138/Copy List with Random Pointer - Leetcode 138.js
/**
 * @param {Node} head
 * @return {Node}
 */
var copyRandomList = function(head) {
    if (!head) return null;

    const oldToNew = new Map();
    let curr = head;

    // First pass: create all nodes and put them in the map
    while (curr) {
        oldToNew.set(curr, new Node(curr.val));
        curr = curr.next;
    }

    // Second pass: set next and random pointers
    curr = head;
    while (curr) {
        const newNode = oldToNew.get(curr);
        newNode.next = oldToNew.get(curr.next) || null;
        newNode.random = oldToNew.get(curr.random) || null;
        curr = curr.next;
    }

    return oldToNew.get(head);
};

/**
 * Definition for a Node.
 * function Node(val, next, random) {
 *     this.val = val;
 *     this.next = next;
 *     this.random = random;
 * }
 */

File: Copy List with Random Pointer - Leetcode 138/Copy List with Random Pointer - Leetcode 138.py
class Solution:
    def copyRandomList(self, head: "Optional [Node]") -> "Optional [Node]":
        if not head:
            return None

        curr = head
        old_to_new = {}

        while curr:
            node = Node(x=curr.val)
            old_to_new[curr] = node
            curr = curr.next

        curr = head

        while curr:
            new_node = old_to_new[curr]
            new_node.next = old_to_new[curr.next] if curr.next else None
            new_node.random = old_to_new[curr.random] if curr.random else None
            curr = curr.next

        return old_to_new[head]

# Time Complexity: O(n)
# Space Complexity: O(n)
